### ISSUE TITLE
Race condition when creating an OpsWorks stack

### ISSUE THREAD

**Issue by charles-at-geospock** [terraform-providers/terraform-provider-aws#1132]
### Terraform Version

(terraspock) Charles-Ferguson-MacBook-Pro:terraspock Charles$ terraform -v
Terraform v0.9.4

(although also tested with v0.10, and the code in question is, seemingly, unchanged)

### Affected Resource(s)

opsworks_stack

### Terraform Configuration Files
Haven't been trimmed to make them the smallest possible, I'm afraid. The debug output should be reasonably clear what has been sent and received.

### Debug Output
Debug output, as discussed below, can be found here:
https://gist.github.com/charles-at-geospock/fda22fd31efb24a22877fc5c93f122a6#file-terraform-debug-section-for-createstack-log

### Expected Behavior
We try to create an OpsWorks stack.
Sometimes it is created but the configuration is wrong.
Stack should have been created with the custom cookbooks and the custom_json.

### Actual Behavior
It hadn't been created with those settings.
The log shows that it should be doing the right thing.
But Amazon is returning as if the change had not happened from beyond the CreateStack operation.
I believe that this is a race condition in the AWS API that is exposed by Terraform performing CreateStack/UpdateStack rapidly after one another.

### References
FIXME - Amazon Forum link.

### Text posted to Amazon forums

Here's the text that I put on the Amazon forums, which will explain this much better.

Hiya,

We use Terraform to build some of our infrastructure, moving away from manually building it, or using python to do so. Whilst testing the creation a part of the infrastructure using OpsWorks I began to notice that some stacks that were built with Terraform were not functional. The instances within the stack had started, but they had not installed any of our software - making them less than useful. This only happened occassionally, and when debugging was enabled in Terraform, it would not reproduce.

The reason that no software had been installed was that:

    - No chef cookbook source was configured
    - The 'use custom cookbooks' setting was 'no'.
    - The 'custom_json' which we expected to be used by our setup wasn't present.

These settings were present in the systems that were working properly, but not the broken one.

This led to the belief that there was a race somewhere within the process of building the stack. Cutting down the code being executed to remove the instances themselves, and any other settings that were not required, in order to reproduce the problem reliably. It could not be reproduced reliably, but it was possible to get debug out from Terraform to see what was happening.

Having traced through the Terraform code itself, it seemed that the stack creation was not parallelised - it would only do the operations in serial, and the SDK code that was invoked appeared to also be serial. Terraform can retry operations, but this was not the case in the logged failures.

Whilst creating the stack, Terraform does the following:

    -   <get parameters for creation>
    -   CreateStack (with minimal required parameters)
    -   UpdateStack
    -   DescribeStack
    -   <set terraform state from described stack>

This allows them to reduce the maintenance of the create/update code to operations which are required for the create, and then modify what's left. This means that there's no need to duplicate some of the setup code in the UpdateStack operation into the CreateStack operation. The two operations are separated (in the common case of not supplying custom security groups or vpc) by an explict 30 second sleep in order to wait on the implicit creation of the security groups which is performed by AWS but not checkable by the API (according to the Terraform source comment).

In our case, this does not apply, so this delay does not happen - the CreateStack is followed immediately by an UpdateStack. In tests, this has been seen to create the stack, returning 200 OK, then update to set a number of settings (including custom_json and cookbooks location), and for this to also be successful. The read back of the settings with DescribeStack then returns and reports the state of the stack - which is missing this cookbook and custom_json setting.

In our case, we do supply our own security groups, so this delay does not happen - the CreateStack is followed immediately by an UpdateStack. In tests, this has been seen to create the stack, returning 200 OK, then update to set a number of settings (including custom_json), and for this to also be successful. The read back of the settings with DescribeStack then returns and reports the state of the stack - which is missing the custom_json setting.

A log of the details from terraform's log contains the following (with some details redacted):

See GitHub Gist: https://gist.github.com/charles-at-geospock/fda22fd31efb24a22877fc5c93f122a6#file-terraform-debug-section-for-createstack-log

In CloudTrail, we see these operations are reported within the same second and actually appear as UpdateStack followed by CreateStack (where, when it works, these happen the other way around, as might be expected more commonly).

So, within Terraform, these operations would not race too often because of the 30 second delay. Within other systems, the configuration would be sent exclusively within the CreateStack operation, and would not be followed by an immediate UpdateStack, or if they did it may be delayed more, so this problem would not be seen.

My belief in the way that this is happening is thus:

....Terraform sends basic CreateStack request.
....Amazon receive the CreateStack request.
........The stack identifier is created.
........Some details are recorded immediately at this point within the stack record.
........Requests are made to set other state up, which is queued by some process. This includes the security group set up that the comments in terraform speak of, and the custom_json that we're using.
........The stack-id is returned (despite the operation not actually being complete yet).
....Terraform sends UpdateStack request, using the stack-id returned.
....Amazon receives UpdateStack request.
........Amazon updates the record of the stack to include the parameters that we have configured, including the custom_json, which is set to the value we sent.
........Success is returned
....Amazon's queued actions for the stack trigger
........The custom_json property, and others are set to the value they were given in the queued request (from CreateStack), where it is empty.
....Terraform sends a DescribeStack request, so that it can read its current state.
....Amazon receives the DescribeStack request
........Amazon returns the current state - which is that custom_json is now empty.
....Terraform stores this as the current state. Because it is the current state - Amazon have lost the updates that were sent between the operations.

How can this be fixed?

1. Simple hack: Update Terraform to make the 30 second sleep apply to all stack creation. Makes the sleep unconditional.

2. Simple hack: Update Terraform to add a 10 second sleep as an 'else' to the security groups check. Makes the data update step slightly slower but reliable (in tests, where there was any delay it seemed to always work)

3. Less simple: Refactor the Terraform create process to be:
........Create basic parameters needed for stack creation.
........Update with the shared parameters that Update would have sent.
........Call CreateStack
........Don't call UpdateStack (because the parameters are already set)
........And the update process would include a call to set up the shared parameters in the same way.
........Makes the code organised along the lines of the Amazon expectations, but doesn't address the security groups creation (which would probably have to remain).

4. Hard: (assuming that this really *is* an API problem) Get Amazon to rework the API so that (one of):
........Create is not asynchronous
........Update and Read are guarandeed to Act As If Create was synchronous (ie, they don't take effect until the queued operations have been effected)
........Add an API to allow checking whether the Stack creation has completed (which could be inserted into the existing code in Terraform between Create and Update and would address the Security Groups issue at the same time).

I'll be raising the same commentary with Terraform, but I mention it here in the hope that someone else might have seen the issue, be able to say that the diagnosis is garbage and point me in another direction, or maybe someone at Amazon might be able to give more comment on the issue.

[[ Sorry about the '.' indentation on this post - seems that leading spaces are just eaten. ]]

**Comment by tracypholmes** [comment]
Thank you for using Terraform and for opening up this question. Issues on GitHub are intended to be related to bugs or feature requests with the provider codebase. Please use https://discuss.hashicorp.com/c/terraform-providers for community discussions, and questions around Terraform.

If you believe that your issue was miscategorized as a question or closed in error, please create a new issue using one of the following provided templates: [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md) or [feature request](https://github.com/terraform-providers/terraform-provider-aws/issues/new?labels=enhancement&template=Feature_Request.md). Please make sure to provide us with the appropriate information so we can best determine how to assist with the given issue.
